### PR TITLE
XLA CustomCall interface for envpool

### PR DESCRIPTION
**Pull request by mavenlin**
## Description
Replace #108
Expose the envs' `send/recv` functions as XLA CustomCalls.

## Motivation and Context

Modern DL frameworks like JAX/Tensorflow allows lowering the neural network part of the loop down to native code. However, currently envpool is pure python, meaning the interaction between the algorithm and the env has to be in python.

With this PR, the neural nets can be jitted together with envpool, which minimizes the dispatching latency.

- [ ] I have raised an issue to propose this change ([required](https://envpool.readthedocs.io/en/latest/pages/contributing.html) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] New environment (non-breaking change which adds 3rd-party environment)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## TODO

- [x] Test case on jitted code.
- [x] Prevent envs that has dynamically shaped states/actions from registering XLA on C++ side.

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://envpool.readthedocs.io/en/latest/pages/contributing.html) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make lint` (**required**)
- [x] I have ensured `make bazel-test` pass. (**required**)

### CODE REVIEW

File: envpool/python/lax.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      return new_handle, self._to(state_list, reset=False, return_info=True)
+
+    def send(
+      handle: jnp.ndarray,
+      action: Union[Dict[str, Any], jnp.ndarray],
+      env_id: Optional[jnp.ndarray] = None
+    ) -> Any:
+      action = self._from(action, env_id)
+      self._check_action(action)
+      print([a.shape for a in action])
```
Comment (Trinkle23897) [inline envpool/python/lax.py:206]:
```suggestion
```

**Review by Trinkle23897 — DISMISSED** [review]
Can we add jit-related doc in README and [docs](https://envpool.readthedocs.io/en/latest/content/python_interface.html)?

File: examples/xla_step.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  # Create an envpool that each step only 4 of 8 result will be out,
+  # and left other "slow step" envs execute at background.
+  env = envpool.make_dm("Pong-v5", num_envs=num_envs, batch_size=batch_size)
+
+  handle, recv, send, step = env.xla()
+
+  def actor_step(iter, loop_var):
+    handle0, states = loop_var
+    action = policy(states.observation.obs)
+    handle1 = send(handle0, action, states.observation.env_id)
```
Comment (mavenlin) [inline examples/xla_step.py:82]: User may easily forget to add `env_id`, currently `full_env_id` will be added. We need to check the `batch_size` vs `num_env` first and raise an exception if necessary.

Comment (mavenlin) [reply examples/xla_step.py:82]: @Trinkle23897

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**